### PR TITLE
Re-implement common interface bindings

### PR DESCRIPTION
**Pull request by PerezHz**
This PR proposes to re-implement the common interface bindings to the `DifferentialEquations.jl` ecosystem essentially by adding `TaylorMethod` as an ODE solver method of `OrdinaryDiffEq.jl` in the sense `TaylorAlgorithm <: OrdinaryDiffEqAdaptiveAlgorithm`. So the proposed approach is to extend `initialize!` and `perform_step!` methods in `OrdinaryDiffEq.jl`, while handling step size control via corresponding method extensions of `stepsize_controller!` and `step_accept_controller!`. By doing this, calling `solve` yields the same results as a direct call to `taylorinteg` with comparable performance, while allowing the use of a larger set of keywords with `TaylorMethod`, including `saveat`, `tstops` and even `callback` (currently, discrete callbacks seem to work okay while continuous callbacks do not).

On the `TaylorIntegration.jl` side, I extended `jetcoeffs!` and `stepsize`, as well as `TaylorSeries.evaluate!` to allow use of high-dimension arrays. Following this strategy, current tests in `master` pass, and also added some more tests. In particular, this PR might help to find a way out of #103 and #109.

There are some issues remaining: besides the continuous callbacks (which I'm currently trying to sort out) I'm not completely sure how to handle the FSAL stuff, but since it is used in 3rd-order Hermite interpolation in `OrdinaryDiffEq`, I added FSAL-related updates in `initialize!` and `perform_step!`, although I'm not sure if I did it right. Also, I left most of the keywords in the warning list `warnkeywords`, although a better approach might now be the inverse: to ditch the keyword warning list altogether (or at least most of it?), and handle any issues that will arise in the future.

Surely there are lots of things that can be improved, so feedback is appreciated!

cc: @ChrisRackauckas @lbenet

### CODE REVIEW

**Comment by PerezHz** [comment]
Updates: I found a way, thanks to @SebastianM-C's work in #108, to finally handle `DynamicalODEProblem`s with TaylorIntegration.jl. A thing to note is that use of `@taylorize` for speedup is currently not allowed for DynamicalODEProblem. Anyway, with the latest changes this PR fixes #109. We can always take care of performance for DynamicalODEProblem, via use of `@taylorize`, in a future PR.

**Comment by lbenet** [comment]
Thanks for the great effort! 

Please ping me, and perhaps @SebastianM-C too, whenever you think this is ready for review.

**Comment by PerezHz** [comment]
> Please ping me, and perhaps @SebastianM-C too, whenever you think this is ready for review.

The latest update is that I think I found a way to make continuous and vector callbacks work (discrete callbacks were already working), which I hope will help users of TaylorIntegration.jl through the common interface.

I'm still not 100% percent sure about the FSAL thing, but things appear to be working well, so... 🤷‍♂️. The keyword warning list perhaps needs to be updated in some way as well, but here's where some feedback might help to know what's the best way to proceed going forward. I think we should add support, respectively, for in-place format in scalar problems and out-of-place format in array problems, but we might tackle that in another PR. Finally, in the tests some funny warnings started to appear:
```
WARNING: Method definition vec(Number) in module FiniteDiff at
C:\Users\appveyor\.julia\packages\FiniteDiff\EBys0\src\jacobians.jl:128 
overwritten in module DiffEqDiffTools at
C:\Users\appveyor\.julia\packages\DiffEqDiffTools\uD0fb\src\jacobians.jl:114.
```
but that seems to be related to FiniteDiff and DiffEqDiffTools, since we're not overwriting any `vec` methods here.

Otherwise, this PR is ready for review!

cc: @lbenet @SebastianM-C

File: src/common.jl [hunk 1]
```diff
@@ -1,194 +1,283 @@
...
-                dd = reshape(du, sizeu); uu = reshape(u, sizeu);
-                f(dd, uu, p, t); u = vec(uu); du = vec(dd); 0)
-        u0 = vec(prob.u0)
-        if isempty(saveat_vec)
-            t, vectimeseries = taylorinteg(f!, u0, prob.tspan[1], prob.tspan[2],
-                alg.order, abstol, prob.p, maxsteps=maxiters, parse_eqs=false)
+    sizeu = size(prob.u0)
+    # This if block handles DynamicalODEProblems
+    # credit to @SebastianM-C for coming up with this solution
+    if prob.f isa DynamicalODEFunction
```
Comment (SebastianM-C) [inline src/common.jl:192]: From what I understand, this special casing is needed because the `jetcoeffs!` function calls the user function with `xaux` which is an `Array`. I was wondering if it would be possible to just use the `DiffEqFunction` interface (i.e. call `prob.f` with `f(du,u,p,t)` and `u` with the same type as `prob.u0`), but I'm not sure I understand how the `jetcoeffs!` code works and if the above suggestion would make sense.

**Review by SebastianM-C — COMMENTED** [review]
Thanks for adding support for `DynamicalODEProblems`. I was thinking if a more general approach where the conversion to `ODEProblem` is no longer required would be possible.

File: src/common.jl (hunk 1 above)
Comment (PerezHz) [reply src/common.jl:192]: Thank you for your review and your suggestions! You're right, it'd be nice if we didn't have to convert everything to an `ODEProblem`... after reading your comment here I gave it a shot, and I think just got to overcome this conversion, at least for the in-place case (I should push that commit in a couple of minutes). The out-of-place version won't work as of now, since `TaylorIntegration.jl` currently does not support oop for `prob.u0 <: AbstractArray`. But once we add that, perhaps in another PR, we'll be able to get rid of most of the `if`s and `else`s, while avoiding the conversion from `DynamicalODEProblem` to `ODEProblem`.

**Comment by PerezHz** [comment]
Just pushed a commit which fixed an error in the casing for out-of-place DynamicalODEProblem (the returned value from f wasn't being passed correctly) and added some tests.

**Comment by PerezHz** [comment]
@SebastianM-C I was thinking... is there a way to convert an out-of-place `DynamicalODEProblem` to an in-place `DynamicalODEProblem`? Since we now can handle in-place `DynamicalODEProblem` if somehow we were able to do that conversion, that way we could avoid the conversion to `ODEProblem` for out-of-place `DynamicalODEProblem`...

**Comment by PerezHz** [comment]
I was trying stuff like
```julia
if prob.f isa DynamicalODEFunction && !isinplace
    f1! = (dv, v, u, p, t) -> (dv .= prob.f.f1(v, u, p, t); 0)
    f2! = (du, v, u, p, t) -> (du .= prob.f.f2(v, u, p, t); 0)
    prob.f = DynamicalODEFunction{true}(f1!, f2!)
    # ...
```
which is similar to what we do for out-of-place `ODEProblem`, but then I get an error:
```
ERROR: MethodError: Cannot `convert` an object of type
  DynamicalODEFunction{true,ODEFunction{true,typeof(iip_ṗ),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},ODEFunction{true,typeof(iip_q̇),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}} to an object of type
  DynamicalODEFunction{false,ODEFunction{false,typeof(oop_ṗ),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},ODEFunction{false,typeof(oop_q̇),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}}
Closest candidates are:
  convert(::Type{T}, ::T) where T at essentials.jl:171
  DynamicalODEFunction{false,ODEFunction{false,typeof(oop_ṗ),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},ODEFunction{false,typeof(oop_q̇),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing}(::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any) where {iip, F1, F2, TMM, Ta, Tt, TJ, JVP, VJP, JP, SP, TW, TWt, TPJ, S, TCV} at /Users/Jorge/.julia/packages/DiffEqBase/3iigH/src/diffeqfunction.jl:64
Stacktrace:
 [1] setproperty!(::ODEProblem{RecursiveArrayTools.ArrayPartition{Float64,Tuple{SArray{Tuple{2},Float64,1,2},SArray{Tuple{2},Float64,1,2}}},Tuple{Float64,Float64},false,DiffEqBase.NullParameters,DynamicalODEFunction{false,ODEFunction{false,typeof(oop_ṗ),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},ODEFunction{false,typeof(oop_q̇),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem}, ::Symbol, ::Function) at ./Base.jl:34
 [2] top-level scope at REPL[35]:1
```

In the tests, if I try

```julia
oop_prob.f = iip_prob.f
```
the same error shows up.

File: test/common.jl [hunk 2]
```diff
@@ -235,11 +241,11 @@ using StaticArrays
         function oop_ṗ(p, q, params, t)
             dp1 = -q[1] * (1 + 2q[2])
             dp2 = -q[2] - (q[1]^2 - q[2]^2)
-            @SVector [dp1, dp2]
```
Comment (SebastianM-C) [inline test/common.jl]: Why did you remove `SVector` initial conditions?

File: src/common.jl [hunk 3]
```diff
@@ -186,40 +186,28 @@ function DiffEqBase.solve(
...
-        u0 = convert(Array{eltype(prob.u0)}, prob.u0)
-        _prob = ODEProblem(f, u0, prob.tspan, prob.p; prob.kwargs...)
-        # DiffEqBase.solve(prob, _alg, args...; kwargs...)
-        integrator = DiffEqBase.__init(_prob, _alg, args...; kwargs...)
-    else
-        f = prob.f
-        if !isinplace && typeof(prob.u0) <: AbstractArray
-            f! = (du, u, p, t) -> (du .= f(u, p, t); 0)
+    f = prob.f
+    if !isinplace && typeof(prob.u0) <: AbstractArray
```
Comment (SebastianM-C) [inline src/common.jl:190]: Is `typeof(prob.u0) <: AbstractArray` necessary?
I think that the only case that we have `prob.u0` not be an array is when it is a scalar differential equation, but in that case it can only be out-of-place, so `!isinplace` is already true.

**Review by SebastianM-C — COMMENTED** [review]
This is a much better solution as it preserves the problem type and avoids converting the user input to an `Array`. This should improve both performance and generality/composability.

File: test/common.jl (hunk 2 above)
Comment (PerezHz) [reply test/common.jl]: Thanks for checking this! I removed them temporarily because it doesn't work with latest commit 😅 since `setindex!` is not defined for `SArray{S,Taylor1}`, I'm trying to fix that right now, just pushed this to see how travis reacted

Comment (PerezHz) [reply test/common.jl]: I'll add them again as soon as I fix this

File: src/common.jl [hunk 4]
```diff
@@ -1,194 +1,258 @@
...
-      push!(tstops_internal, tdir_tf)
+    if verbose
+        warned = !isempty(kwargs) && check_keywords(alg, kwargs, warnlist)
+        warned && warn_compat()
+    end
+
+    f = prob.f
+    if !isinplace && typeof(prob.u0) <: AbstractArray
+        if prob.f isa DynamicalODEFunction
+            f1! = (dv, v, u, p, t) -> (dv .= prob.f.f1(v, u, p, t); 0)
```
Comment (SebastianM-C) [inline src/common.jl:192]: Since we are in the out-of-place case, shouldn't this _not_ use broadcasting to modify the elements of `dv` in-place? 
```suggestion
            f1! = (dv, v, u, p, t) -> (dv = prob.f.f1(v, u, p, t); 0)
```

File: src/common.jl [hunk 5]
```diff
@@ -1,194 +1,258 @@
...
+            f2! = (du, v, u, p, t) -> (du .= prob.f.f2(v, u, p, t); 0)
```
Comment (SebastianM-C) [inline src/common.jl:193]: Since we are in the out-of-place case, shouldn't this _not_ use broadcasting to modify the elements of `dv` in-place?
```suggestion
            f1! = (dv, v, u, p, t) -> (dv = prob.f.f1(v, u, p, t); 0)
            f2! = (du, v, u, p, t) -> (du = prob.f.f2(v, u, p, t); 0)
```

File: src/common.jl (hunk 4 above)
Comment (SebastianM-C) [inline src/common.jl:192]: This is what causes test to fail in the OOP case with immutable types.

**Review by SebastianM-C — COMMENTED** [review]
The translation of out-of-place problems to in-place ones is fundamentally incompatible with immutable types. From what I understand from the DiffEq code, the OOP code path is made specifically for immutable types.

File: src/common.jl (hunk 3 above)
Comment (PerezHz) [reply src/common.jl:190]: Thank you for your suggestion. If we think an oop scalar problem, and if we removed that `typeof(prob.u0) <: AbstractArray`, then in that case the code would try to transform it into a in-place scalar problem, but currently TaylorIntegration does not support in-place scalar problems; that's why I left it there. Tests actually break if I remove that part

File: test/common.jl (hunk 2 above)
Comment (PerezHz) [reply test/common.jl]: They are back now... not the prettiest solution, but it was the best compromise until we properly support oop problems in TaylorIntegration.jl

File: src/common.jl (hunk 4 above)
Comment (PerezHz) [reply src/common.jl:194]: Good catch, thanks! I'm proposing, as a temporary solution for this specific case, to convert the initial condition `prob.u0` from an immutable to a mutable array type (see below)

**Comment by PerezHz** [comment]
Ok, so I think I found a compromise: I found a way to make things to work with `SVectors` when dealing with oop `DynamicalODEProblem` by transforming specifically for this case the initial condition `prob.u0` from an immutable array type to a mutable one (it's highly non-elegant, but does the job). I fully agree that this does not solve the incompatibility of the oop->iip translation with immutable arrays, but solving that deeper issue requires some changes under the hood of TaylorIntegration.jl itself, which I believe we should think through and tackle in another PR.

File: .gitignore [hunk 6]
```diff
@@ -3,3 +3,4 @@ examples/.ipynb_checkpoints
 examples/JuliaCon2017/.ipynb_checkpoints/
 docs/build/
 docs/site/
+.vscode/
```
Comment (lbenet) [inline .gitignore:6]: This may need  a new line (it's the end of the file); this is noted by github.

File: src/common.jl [hunk 7]
```diff
@@ -1,194 +1,267 @@
...
+alg_order, alg_cache, initialize!, perform_step!, @unpack,
+@cache, stepsize_controller!, step_accept_controller!
+
+# TODO: check which keywords work fine
+const warnkeywords = (:save_idxs, :d_discontinuities, :unstable_check, :save_everystep,
+:save_end, :initialize_save, :adaptive, :dt, :reltol, :dtmax,
+:dtmin, :force_dtmin, :internalnorm, :gamma, :beta1, :beta2,
+:qmax, :qmin, :qsteady_min, :qsteady_max, :qoldinit, :failfactor,
+:isoutofdomain, :unstable_check,
+:calck, :progress, :timeseries_steps, :dense)
```
Comment (lbenet) [inline src/common.jl:18]: Can you include some indentation; makes nicer reading the code.

Can you open an issue to  have present the TO DO indicated here?

File: src/common.jl [hunk 8]
```diff
@@ -1,194 +1,267 @@
...
+abstract type TaylorAlgorithm <: OrdinaryDiffEqAdaptiveAlgorithm end
 struct TaylorMethod <: TaylorAlgorithm
     order::Int
 end
+struct _TaylorMethod <: TaylorAlgorithm
+    order::Int
+    parse_eqs::Bool
+end
+
+_TaylorMethod(order; parse_eqs=true) = _TaylorMethod(order, parse_eqs) # set `parse_eqs` to `true` by default
```
Comment (lbenet) [inline src/common.jl:33]: Naive question: why do we need two (essentially identical) structs, instead  of only one? Perhaps by adjusting some constructors, everything works with only one...

File: src/common.jl [hunk 9]
```diff
@@ -1,194 +1,267 @@
...
+
+alg_order(alg::TaylorMethod) = alg.order
+alg_order(alg::_TaylorMethod) = alg.order
+
+# Regarding `isfsal`, thought about setting isfsal to false, but low order
+# interpolation may be helpful (since Taylor interpolation is memory-intensive),
+# so let's set isfsal to true for now, which is the default, so no method
+# definition needed here, so we in principle should support FSAL
+# isfsal(::TaylorMethod) = true # see discussion above
```
Comment (lbenet) [inline src/common.jl:42]: Yet, this line is commented... Is it needed? We can certainly have it there commented, but then clarifying a bit would be very enlightening.

File: src/common.jl [hunk 10]
```diff
@@ -1,194 +1,267 @@
...
+#   uprev::uType
+#   tmp::uType
+#   k::rateType
+#   fsalfirst::rateType
+#   tT::tTType
+#   uT::uTType
+#   duT::uTType
+#   uauxT::uTType
+#   parse_eqs::Bool
+# end) )
```
Comment (lbenet) [inline src/common.jl:80]: Do we need this? Seems it is a left-over of something you were testing...

File: src/common.jl [hunk 11]
```diff
@@ -1,194 +1,267 @@
...
+    if length(k)<2 || always_calc_begin
+        if typeof(cache) <: OrdinaryDiffEqMutableCache
+        rtmp = similar(u, eltype(eltype(k)))
+        f(rtmp,uprev,p,t)
+        copyat_or_push!(k,1,rtmp)
+        f(rtmp,u,p,t+dt)
+        copyat_or_push!(k,2,rtmp)
+        else
+        copyat_or_push!(k,1,f(uprev,p,t))
+        copyat_or_push!(k,2,f(u,p,t+dt))
```
Comment (lbenet) [inline src/common.jl:249]: Can you include some indentation here?

File: src/common.jl [hunk 12]
```diff
@@ -1,194 +1,267 @@
...
+function evaluate!(x::AbstractArray{Taylor1{T},N}, δt::S,
+        x0::AbstractArray{T,N}) where {T<:Number, S<:Number, N}
 
-    tstops_internal, saveat_internal, d_discontinuities_internal
+    # @assert length(x) == length(x0)
+    @inbounds for i in eachindex(x, x0)
+        x0[i] = evaluate( x[i], δt )
+    end
+    nothing
 end
```
Comment (lbenet) [inline src/common.jl:267]: Why not  adding it directly to TaylorSeries?

File: test/common.jl [hunk 13]
```diff
@@ -1,16 +1,24 @@
-using TaylorIntegration, Test, DiffEqBase
+using Test, DiffEqBase, TaylorIntegration
```
Comment (lbenet) [inline test/common.jl:1]: Does the order matter here? If so, i think it is worth adding  a couple of sentences to the docs.

File: src/common.jl [hunk 14]
```diff
@@ -1,194 +1,267 @@
...
+        _prob = prob
+    end
 
-      for t in d_discontinuities
-        tdir_t = tdir * t
-        tdir_t0 < tdir_t ≤ tdir_tf && push!(tstops_internal, tdir_t)
-      end
+    # DiffEqBase.solve(prob, _alg, args...; kwargs...)
+    integrator = DiffEqBase.__init(_prob, _alg, args...; kwargs...)
+    integrator.dt = stepsize(integrator.cache.uT, integrator.opts.abstol) # override handle_dt! setting of initial dt
```
Comment (lbenet) [inline src/common.jl:220]: Previously, there was a default value set for `abstol`; I cannot see it now. So the question is if the user must now to specify it, or is it set somewhere else?

**Review by lbenet — APPROVED** [review]
Thanks for the tremendous work  that there is in this PR! I have left mostly trivial remarks  to clarify a couple of points. 

LGTM and I am in favor of merging this, though maybe we may wait for @SebastianM-C's opinion.

File: .gitignore (hunk 6 above)
Comment (PerezHz) [reply .gitignore]: Fixed, thanks!

File: src/common.jl (hunk 7 above)
Comment (PerezHz) [reply src/common.jl]: Thanks, added indentation and also opened the issue corresponding to this TODO

File: src/common.jl (hunk 8 above)
Comment (PerezHz) [reply src/common.jl]: Good catch! Yes, it was redundant to have those two structures, so I removed one of them

File: src/common.jl (hunk 9 above)
Comment (PerezHz) [reply src/common.jl]: It was more of a personal reminder when trying to figure the internals of DiffEqBase and OrdinaryDiffEq; I've now removed these commented lines

File: src/common.jl (hunk 10 above)
Comment (PerezHz) [reply src/common.jl]: 😅 sorry, thanks! Yes, this was a leftover, it's now gone

File: src/common.jl (hunk 11 above)
Comment (PerezHz) [reply src/common.jl]: Thanks, done!

File: src/common.jl (hunk 12 above)
Comment (PerezHz) [reply src/common.jl:246]: This is the only point which I haven't addressed yet; I will open the corresponding PR in TaylorSeries, and it will perhaps require a new TaylorSeries release

File: test/common.jl (hunk 13 above)
Comment (PerezHz) [reply test/common.jl]: The order makes no difference, I just forgot to leave that line untouched. I've now gone back to the original order, so that this line doesn't show up when doing diffs in the future

File: src/common.jl (hunk 14 above)
Comment (PerezHz) [reply src/common.jl:198]: Good catch! If the user doesn't specify `abstol` it is set automatically by `DiffEqBase.__init` to `abstol=1e-6`

**Comment by PerezHz** [comment]
> Thanks for the tremendous work that there is in this PR! I have left mostly trivial remarks to clarify a couple of points.

Thanks @lbenet for your review! I addressed all your comments except one, which will have to wait a bit while we add the new `evaluate!`method to TaylorSeries. In the meantime, please let me know if there's something that you think can be further improved.
 
> LGTM and I am in favor of merging this, though maybe we may wait for @SebastianM-C's opinion.

I agree we should wait to hear Sebastian's opinion.

File: src/common.jl (hunk 12 above)
Comment (PerezHz) [reply src/common.jl:246]: Update: just opened the corresponding PR. See: https://github.com/JuliaDiff/TaylorSeries.jl/pull/252

**Review by SebastianM-C — APPROVED** [review]

**Comment by lbenet** [comment]
Tests are passing, except in nightly; I am just including the (first) problem spotted by travis [here](https://travis-ci.org/github/PerezHz/TaylorIntegration.jl/jobs/747480423#L598) to address it later.

**Comment by lbenet** [comment]
Shall I go ahead and merge?

**Comment by PerezHz** [comment]
> Shall I go ahead and merge?

Sorry for the late reply, sure, I agree that we should merge!

**Comment by lbenet** [comment]
Merged! Thanks a lot!
